### ISSUE TITLE
Seccomp at POD level always requires too many syscalls

### ISSUE THREAD

**Issue by pjbgf** [kubernetes/kubernetes#84623]
**What happened**:

For a single container pod with `AllowPrivilegeEscalation=false`, setting the seccomp profile at pod or container level results in vastly different syscalls to be whitelisted. 

**What you expected to happen**:

The same (or similar) pod seccomp profile should work at both container or pod levels. Instead, setting profiles at pod level requires excessive permissions, including: `capset`, `set_tid_address`, `setgid`, `setgroups` and `setuid`.

The table below shows what syscalls must be whitelisted to run a simple pod depending on whether the seccomp profile is set at pod or container level:

| Seccomp set at | Whitelisted syscalls |
|------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Container Level | arch_prctl<br>brk<br>close<br>execve<br>exit_group<br>futex<br>mprotect<br>nanosleep<br>stat<br>write | 
| Pod Level | **access**<br>arch_prctl<br>brk<br>**capget**<br>**capset**<br>**chdir**<br>close<br>**epoll_ctl**<br>**epoll_pwait**<br>execve<br>exit_group<br>**fchown**<br>**fcntl**<br>**fstat**<br>**fstatfs**<br>futex<br>**getcwd**<br>**getdents64**<br>**getgid**<br>getpid<br>**getppid**<br>**getuid**<br>**lstat**<br>mprotect<br>nanosleep<br>**newfstatat**<br>**openat**<br>**pause**<br>**prctl**<br>**read**<br>**readlink**<br>**rt_sigaction**<br>**rt_sigprocmask**<br>**set_tid_address**<br>**setgid**<br>**setgroups**<br>**setuid**<br>**statfs**<br>stat<br>**uname**<br>**wait4**<br>write |

The reason for this behaviour seems to be that when seccomp profiles are applied at pod level it is applied to all containers, including the "pause" container. The problem is that this container always run with `AllowPrivilegeEscalation=true`, which makes the container runtime (docker at least) to apply the seccomp early in the process leading to an increased profile footprint. 

Forcing the "pause" container to run with `AllowPrivilegeEscalation=false` (through changes on the source code) results in a lighter seccomp profile:

arch_prctl<br>brk<br>close<br>execve<br>exit_group<br>futex<br>mprotect<br>nanosleep<br>stat<br>write<br>**uname**<br>**pause**<br>**rt_sigaction**

However, refraining from applying the seccomp profile to the "pause" container allows for the exact same profile to work at both levels.

Would there be a problem to either:

1) Change the pause container to not apply the seccomp profile (preferable)?
2) Allow a setting for the pause container to have its `AllowPrivilegeEscalation` configurable?

**How to reproduce it**:

1. Spin up a kubernetes cluster with `hack/local-up-cluster.sh`
2. Place the contents of both `*.json` files into `/var/lib/kubelet/seccomp/`
3. Create the pods defined on `pods.yaml` file
4. Confirm that only the pod pod-level-container-perms is **not** running
```console
$ kubectl get pods -n default
NAME                        READY   STATUS              RESTARTS   AGE
container-level             1/1     Running             0          7s
pod-level-container-perms   0/1     ContainerCreating   0          7s
pod-level-excessive-perms   1/1     Running             0          7s
```

#### Example Files

**pods.yaml**:
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: container-level
  annotations:
    container.seccomp.security.alpha.kubernetes.io/container-level: 'localhost/container-profile.json'
spec:
  containers:
  - name: container-level
    image: alpine
    command: ["sh"]
    args: ["-c", "echo hi && sleep 120"]
    securityContext:
      allowPrivilegeEscalation: false
---
apiVersion: v1
kind: Pod
metadata:
  name: pod-level-excessive-perms
  annotations:
    seccomp.security.alpha.kubernetes.io/pod: 'localhost/pod-profile.json'
spec:
  containers:
  - name: pod-level-excessive-perms
    image: alpine
    command: ["sh"]
    args: ["-c", "echo hi && sleep 120"]
    securityContext:
      allowPrivilegeEscalation: false
---
apiVersion: v1
kind: Pod
metadata:
  name: pod-level-container-perms
  annotations:
    seccomp.security.alpha.kubernetes.io/pod: 'localhost/container-profile.json'
spec:
  containers:
  - name: pod-level-container-perms
    image: alpine
    command: ["sh"]
    args: ["-c", "echo hi && sleep 120"]
    securityContext:
      allowPrivilegeEscalation: false
```

**container-profile.json**
```json
{
    "defaultAction": "SCMP_ACT_ERRNO",
    "architectures": [
        "SCMP_ARCH_X86_64",
        "SCMP_ARCH_X86",
        "SCMP_ARCH_X32"
    ],
    "syscalls": [
        {
            "names": [
                "arch_prctl",
                "brk",
                "close",
                "execve",
                "exit_group",
                "futex",
                "mprotect",
                "nanosleep",
                "stat",
                "write"
            ],
            "action": "SCMP_ACT_ALLOW"
        }
    ]
}
```

**pod-profile.json**
```json
{
    "defaultAction": "SCMP_ACT_ERRNO",
    "architectures": [
        "SCMP_ARCH_X86_64",
        "SCMP_ARCH_X86",
        "SCMP_ARCH_X32"
    ],
    "syscalls": [
        {
            "names": [
                "arch_prctl",
                "brk",
                "capget",
                "capset",
                "chdir",
                "close",
                "execve",
                "exit_group",
                "fstat",
                "fstatfs",
                "futex",
                "getdents64",
                "getppid",
                "lstat",
                "mprotect",
                "nanosleep",
                "newfstatat",
                "openat",
                "prctl",
                "read",
                "rt_sigaction",
                "statfs",
                "setgid",
                "setgroups",
                "setuid",
                "stat",
                "uname",
                "write"
            ],
            "action": "SCMP_ACT_ALLOW"
        }
    ]
}
```

**Anything else we need to know?**:

This can be reproduced with both dockershim and cri-o runtime.

**Environment**:
runC and kubelet were the latest from master:

```
$ runc -v [HEAD]
runc version spec: 1.0.1-dev

$ kubelet --version 
Kubernetes v1.18.0-alpha.0.642+0968636760dae2

$ docker version
Client:
 Version:           18.09.7
 API version:       1.39
 Go version:        go1.10.1
 Git commit:        2d0083d
 Built:             Fri Aug 16 14:20:06 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.09.7
  API version:      1.39 (minimum version 1.12)
  Go version:       go1.10.1
  Git commit:       2d0083d
  Built:            Wed Aug 14 19:41:23 2019
  OS/Arch:          linux/amd64
  Experimental:     false

$ crio -h
VERSION:
   1.17.0-dev
commit: "6ea0e8c807d52a30863f8f8a05272e0d4028c7f9"
```

/area security
/cc @tallclair 
/assign @pjbgf 
/sig node

**Comment by pjbgf** [comment]
Retesting this I noticed that actually it seems to be specific to dockershim. On a clean environment this was not reproduceable with CRI-O.

**Comment by pjbgf** [comment]
On a fresh install the issue is reproducible in both dockershim and cri-o.

**Comment by Random-Liu** [comment]
> Forcing the "pause" container to run with AllowPrivilegeEscalation=false (through changes on the source code) results in a lighter seccomp profile

I think we can always set `AllowPrivilegeEscalation=false` for pause container.

And for pause container, I think it makes sense to set a default seccomp profile inside the container runtime? It seems that the seccomp profile for pause container shouldn't be configurable by users, because the behavior is fixed for a dedicated pause container image, we know exactly what it needs.

@pjbgf Hm, why `AllowPrivilegeEscalation` would change the syscall requirement?

**Comment by pjbgf** [comment]
That's awesome, @Random-Liu, thank you. I will give it a go on a PR for that.

> @pjbgf Can you explain why `AllowPrivilegeEscalation` would change the syscall requirement? Not sure how it works.

Historically, Docker always applied the seccomp profile very early in the process of starting  containers, which meant that things such as drop caps and setuid had to be included in the users seccomp profile. From version 1.12, they added the option of executing it as the very last thing just before the `exec` when `AllowPrivilegeEscalation=true`.

In runC, when `AllowPrivilegeEscalation=true`, seccomp is applied before [finalizeNamespace](https://github.com/opencontainers/runc/blob/b133feaeeb2e69ba94aa95eac3a455a143435ea9/libcontainer/init_linux.go#L122) which handles capabilities, setuid/setgid, etc:
https://github.com/opencontainers/runc/blob/b133feaeeb2e69ba94aa95eac3a455a143435ea9/libcontainer/standard_init_linux.go#L156

Otherwise it happens here:
https://github.com/opencontainers/runc/blob/b133feaeeb2e69ba94aa95eac3a455a143435ea9/libcontainer/standard_init_linux.go#L205

**Comment by pjbgf** [comment]
This is now resolved as of v1.19.
